### DISCUSSION TITLE
[Feature Request] Add/Append the same clip to the start AND/OR the end of multiple files in Batch

### DISCUSSION THREAD

**Discussion by LeonCheung** [mifi/lossless-cut#1975] [Ideas]
Hi mifi, first I wanna say a big big thanks to you for developing this great tool. It's very handy and working really like magic!

Allow me to propose a new feature (if not implemented yet), which is just as the title says. For example:

I have lots of media files, say, for a TV or TikTok video series, which need to add a short clip to the start/ end, or add two clips  respectively for start and end, normally it would be very short one, like a 2~3 seconds, typically as Intro / Outro.

Files may look ike this:

```
intro.mp4
outro.mp4
01.mp4
02.mp4
.
.
.
80.mp4
```

Now I want to add start.mp4 to every file listed (from 01.mp4 to 80,np4). Also, I would like to do the same with end.mp4 to each file as an ending.

So, the logic could be very easy to understand like this:

```
output_for_01.mp4 = intro.mp4 + 01.mp4 + outro.mp4
output_for_02.mp4 = intro.mp4 + 02.mp4 + outro.mp4
.
.
.
output_for_80.mp4 = intro.mp4 + 80.mp4 + outro.mp4
```

It's useful for YouTubers or TikTokers to add the same cover clip AND/OR end clip to a bunch of files. Here is a humble UI prototype for inspiration:

![suggestion-for-LosslessCut](https://github.com/mifi/lossless-cut/assets/494788/be595c24-c08f-48c7-91d0-a03ae0159ad8)

Thanks again for creating this great tool.  Cheers. 🍻

**Comment by mifi** [comment]
Hi. Thanks for your idea, but unfortunately I think it's not very practical, because when losslessly concatenating files, **all** the files need to be of the exact same resolution, fps, codec, timebase, colorspace and similar bitrate. I think often when people make a title/intro and outro screen the codec parameters will often be different from the actual main footage, and then it's going to fail.

**Reply by LeonCheung** [reply]
@mifi Hi, thanks for your reply. ❤️ 

**A shorter reply:**
Isn't that the same requirement for the current Merge tool? 😂 I think it's simply a request for a Batch Mode to dramatically save time from the repetitive work of merging each episode one by one. (same like [this](https://www.youtube.com/watch?v=xkHJoq47VEw&t=53s))

**A longer reply:**
Yes, it will fail if they don't know what they do. However, for those who always follow the rule of the existing Merge tool, no problem would occur, right? And we could always pop up instructions on how to avoid failure and show the error message if there's any. (as LosslessCut always does)

I mean, when the intro/outro uses the **same profile** as the same series, then no problem would occur for sure. Just imagine how to add the same intro and/or outro for a series of 100 episodes with LosslessCut ONE BY ONE **without a batch feature**. Trust me, it's horrible and really time-consuming because that is what I'm currently doing. 😂

BTW, I sent 300 CNY via WeChat but it told me the receiptor is... Stripe? Just want to make sure you will ultimately receive my donation to your great work.  :)

**Reply by mifi** [reply]
> Isn't that the same requirement for the current Merge tool? 😂 I think it's simply a request for a Batch Mode to dramatically save time from the repetitive work of merging each episode one by one. (same like [this](https://www.youtube.com/watch?v=xkHJoq47VEw&t=53s))

that's correct.

It's the first time I heard this feature request so it seems a bit niche to me so far, but if it's something that many people want implmented I will definitely consider it.

If I understand your request correctly, the same could be achieved using simple bash scripting, something like this(with help from ChatGPT, not tested!):

```bash
#!/bin/bash

for file in input/*.mp4; do
  # Create a temporary file listing all the files to concatenate
  echo "file 'intro.mp4'" > concat-list.txt
  echo "file '$file'" >> concat-list.txt
  echo "file 'outro.mp4'" >> concat-list.txt
  
  # Use FFmpeg to concatenate the files
  ffmpeg -f concat -safe 0 -i concat-list.txt -c copy "$file-output.mp4"
  
  # Remove the temporary file
  rm concat-list.txt
done
```

here's my prompt: "give me a bash script that losslessly concatenates three files using the concat demuxer: intro.mp4 $file outro.mp4"

Using the above script, you don't even have to click anything at all.

> BTW, I sent 300 CNY via WeChat but it told me the receiptor is... Stripe? Just want to make sure you will ultimately receive my donation to your great work. :)

was it on apr 27? if so, then I received, thanks!

**Reply by LeonCheung** [reply]
Thanks mifi, but sorry, I know little about coding... how should I use or run this script, please? A few steps would be really helpful, Thanks!

This is what I tried so far by googling:

![image](https://github.com/mifi/lossless-cut/assets/494788/fadc0557-adfb-4e72-b5a4-6ee7600e79ad)

**Reply by mifi** [reply]
oh, the script was written for Bash (UNIX), not Windows Batch. I think you'd needs something like WSL to run bash on windows, but i'm not experienced with that. or re-write the script to Windows Batch, but I'm not experienced with that either. maybe ChatGPT can do it...

**Reply by LeonCheung** [reply]
Okay, thanks mifi, I'll give it a try later. Meanwhile, I'll keep an eye on this and see if it's worth any official implementation here in the future.  :)
